source/common/router/router.cc

```diff
@@ -514,45 +540,112 @@ void Filter::onRequestComplete() {
   }
 }
 
-void Filter::onDestroy() {
-  if (upstream_request_ && !attempting_internal_redirect_with_complete_stream_) {
-    upstream_request_->resetStream();
-  }
-  cleanup();
-}
+void Filter::onDestroy() { cleanup(); }
 
 void Filter::onResponseTimeout() {
   ENVOY_STREAM_LOG(debug, "upstream timeout", *callbacks_);
-  cluster_->stats().upstream_rq_timeout_.inc();
 
-  // It's possible to timeout during a retry backoff delay when we have no upstream request.
-  if (upstream_request_) {
-    if (upstream_request_->upstream_host_) {
-      upstream_request_->upstream_host_->stats().rq_timeout_.inc();
+  // Reset any upstream requests that are still in flight.
+  for (auto it = upstream_requests_.cbegin(); it != upstream_requests_.cend(); it++) {
+    UpstreamRequest* upstream_request = it->get();
+    // Don't record a timeout for upstream requests we've already seen headers
+    // for.
+    if (!upstream_request->upstream_headers_) {
+      cluster_->stats().upstream_rq_timeout_.inc();
+      if (upstream_request->upstream_host_) {
+        upstream_request->upstream_host_->stats().rq_timeout_.inc();
+      }
+
+      // If this upstream request already hit a "soft" timeout, then it
+      // already recorded a timeout into outlier detection. Don't do it again.
+      if (!upstream_request->outlier_detection_timeout_recorded_) {
+        updateOutlierDetection(timeout_response_code_, upstream_request);
+      }
+      upstream_request->resetStream();
+
+      chargeUpstreamAbort(timeout_response_code_, false, upstream_request);
     }
-    upstream_request_->resetStream();
   }
 
-  updateOutlierDetection(timeout_response_code_);
   onUpstreamTimeoutAbort(StreamInfo::ResponseFlag::UpstreamRequestTimeout);
 }
 
-void Filter::onPerTryTimeout() {
-  updateOutlierDetection(timeout_response_code_);
+// Called when the per try timeout is hit but we didn't reset the request
+// (hedge_on_per_try_timeout enabled).
+void Filter::onSoftPerTryTimeout(UpstreamRequest* upstream_request) {
```

Comment:
Since this is never null from what I can tell it would probably be better to pass this by reference, same with other places where you pass a pointer